# README #

### What is this repository for? ###

This is the repository of IRIC-Portal Web interface. The application is written in Java, using Hibernate for ORM, Spring to manage the beans and configurations and ZK for the viewers. There are separate repositories for the backend Oracle and data generation pipeline scripts.

* Version 1
	* New Features
	* Feature updates
	* Bug fixed
	* Updates are logged in [http://oryzasnp.org/iric-portal/_about.zul](http://oryzasnp.org/iric-portal/_about.zul "updates")


### How do I get set up? ###

* Setup  
	
	- Download the HDF5 data files, Blast database files and save them in a dedicated directory in the web server.
	- Download the customized JBrowse from . Extract it in the webserver and create a symbolic link (ex. jbrowse) from the tomcat webapps directory pointing to the extracted JBrowse directory.
	- Download the project and open in Eclipse (or preferably MyEclipse). 
	- Set the database connection parameters in bean IRIC\_ProductionDS in /resources/iric\_prod\_crud-dao-context.xml
	- Set the parameters defined in the static class org.irri.iric.portal.AppContext, specially:
	
			getFlatfilesDir()	 	Directory of SNP-Seek data files in the server (using server file system) 
			getHostname()			Web server hostname or IP address 
			getTempDir()			Directory to write temp files accessible to the internet, but using the server file system (ex. /path/to/tomcat/webapps/temp)
			getJbrowseDir() 		JBrowse folder name as deployed in web server (ex. jbrowse)
		
		These configurations are currently hard-coded, but will be made configurable through XML later. 

	- Compile the application and export into a war file. (ex. iric-portal.war)

	- Download and install the HDF5 library from [https://www.hdfgroup.org/HDF5/release/obtain5.html](https://www.hdfgroup.org/HDF5/release/obtain5.html "HDF5 library") into the web server (also available in ). Add the HDF5 library installation directory to Tomcat. In /path/to/tomcat/bin/setenv.sh add the line
	
			export JAVA_OPTS="-Djava.library.path=/path/to/hdf5/lib"

	
	- Deploy the generated war in Tomcat
	- Set a cron job to regularly empty the directory defined in getTempDir() 
	
  
* Dependencies    
	All required jar files are included in the project in /WebRoot/WEB-INF/lib

* Database configuration  
	To connect to the database, edit the username, password and url for the Data Source Setup in the file /resources/iric\_prod\_crud-dao-context.xml. Please contact us for these values, with your development machine IP address.

		<class="org.apache.commons.dbcp.BasicDataSource" destroy-method="close" name="IRIC_ProductionDS">
			<property name="driverClassName" value="oracle.jdbc.driver.OracleDriver"/>
			<property name="username" value="*"/>
			<property name="password" value="*"/>
			<property name="url" value="jdbc:oracle:thin:@*"/>
			<property name="maxIdle" value="16"/>
			<property name="maxActive" value="32"/>
		</bean>


### Project organization ###

The project files are organized in these folders and packages. Javadoc documentation is also available in [http://oryzasnp.org/snpseek-javadoc/](http://oryzasnp.org/snpseek-javadoc/)

* /src   
	the source code	organized into java packages described below. The packages are organized as descibed with the UML class diagrams in /uml
* /src/applicationContext-business.xml    
	The configuration files defined in /resources are imported in sequence as defined in this file. When a new package is created, it should be registered in this file and in this order: 
		
		Domains, DAOs, Service, Web or WS or ZK Controllers
to makes sure that the Spring dependency injection works without NULL reference assignments
* /resources    
	configuration files.
		
		iric_prod_crud-dao-context.xml
		iric_prod_crud-security-context.xml
		iric_prod_crud-service-context.xml
		iric_prod_crud-web-context.xml
* /uml
	UML class diagram files
* /doc	
	source code documentation generated by Javadoc
* /WebRoot/WEB-INF/applicationContext-security.xml      
	Spring-security configuration
* /WebRoot   
	Root directory of the web application, as viewed in the browser. We are using ZK for the viewers and the webpages are defined in *zul files, instead of html. The most important files are:

		genotype.zul	SNP query page
		variety.zul		Variety query page
		locus.zul		Gene loci query page
		download.zul	Download page
		myworkspace.zul	My List page
		gwas.zul		GWAS analysis page
		home.zul		Front page
	

	The web layout is defined by these files
		
		index.zul		Landing/default page
		template.zul	Page layout
		banner.zul		Banner and menu page
		footer.zul		Footer page	
		_*				files starting with _ are the corresponding containers to apply the template to the pages
	
* README.md   this file
	




### Source code organization ###


The Java source codes are organized into these major packages  

		org.irri.iric.portal.domain		domain objects interface, used by all services
		org.irri.iric.portal.dao		data access object interface, used by all services
		org.irri.iric.portal.genotype	SNPs and Indels queries
		org.irri.iric.portal.variety	variety phenotype and passport data queries
		org.irri.iric.portal.genomics	gene loci, gene ontology queries
		org.irri.iric.portal.gwas		GWAS analysis modules 
		org.irri.iric.portal.chado.oracle		DAO and Domain implementation for data stored in Oracle using Chado schema
		org.irri.iric.portal.hdf5		DAO and Domain implementation for data stored in HDF5 format
		org.irri.iric.portal.ws			web services
		org.irri.iric.portal.admin		manages workspace, users


* Service

The web application follows the Model-View-Controller (MVC) architecture. There are currently four major functionalities of the application: genotype query, variety query, gene loci query and user workspace management, which is organized in:

		org.irri.iric.portal.genotype
		org.irri.iric.portal.variety
		org.irri.iric.portal.genomics
		org.irri.iric.portal.admin

Within these packages are  **.service**  and **.zkui** subpakages. Each service has a facade interface, which serves as an entry point for the domain. To make the user interfaces, web interfaces and WS-API independent of the actual data sources, the viewers, controllers or web services should use only these Facade interfaces:

		org.irri.iric.portal.genotype.service.GenotypeFacade
		org.irri.iric.portal.variety.service.VarietyFacade
		org.irri.iric.portal.genomics.service.GenomicsFacade
		org.irri.iric.portal.admin.WorkspaceFacade

Each method of these facade call other **Service** objects, which implements the query logic. The service objects then use **DAO** or data access object interfaces to query the database.  

* Views

Our web interface use the ZK framework [http://www.zkoss.org/](http://www.zkoss.org/ "ZK framework"). The view controllers are in the **org.irri.iric.portal.\*.zkui** packages which include the ZK view controller and supporting classes for the components (buttons, listboxes, tables, checkboxes, etc.) in the user interface. The interface termplate itself is defined in the *.zul files in /WebRoot
 

* Web service

The web services API are implemented using Spring REST Controller to map URL paths to service and  facade methods. There are currently three major web services:

		org.irri.iric.portal.ws.rest.GenotypeWS		For methods in *GenotypeFacade
		org.irri.iric.portal.ws.rest.VarietyWS 		For methods in *VarietyFacade
		org.irri.iric.portal.ws.rest.BlastWS		For methods in *LocalAlignmentService

To hide or add more attributes, or change data types to the response objects, WS-specific entities are implemented to wrap the DAO-generated entities. These web service response objects are defined in **org.irri.iric.portal.ws.entity**

The WADL file [http://oryzasnp.org/iric-portal/ws/application.wadl](http://oryzasnp.org/iric-portal/ws/application.wadl "WADL") which can be used to generate client objects, is generated automatically by Spring from the defined mappings. This file is also used to generate the API documentation. The file /WebRoot/api-docs/application-api-docs.json is such documentation in Swagger-json format, can be viewed by swagger-ui in [http://oryzasnp.org/iric-portal/swagger-ui/index.html](http://oryzasnp.org/iric-portal/swagger-ui/index.html "swagger-ui")   


* Data stores

Every data store should implement the Domain and DAO interfaces for each data it can provide. These are the currently implemented data stores:

		org.irri.iric.portal.chado.*	To access Oracle with Chado schema
		org.irri.iric.portal.hdf5.*		To read HDF5 genotype files
		org.irri.iric.portal.flatfile.*	To read ASCII genotype files
	
Within these packages are subpackages:

		*dao 		where the actual data retrieval is implemented, like SQL query or file reading
		*domain 	holds the entity attribute values
	
* Utility classes

These are classes that are frequently used by all the modules.

		org.irri.iric.portal.AppContext	This class provides application-wide access to the Spring ApplicationContext. 
										It is used to define global application parameters. These parameters are currently hard-coded,
										but has the capability to be overriden with *.xml defined values.
		org.irri.iric.portal.dao.ListItemsDAO, org.irri.iric.portal.dao.ListItemsDAOImpl
										Iterface and an implementation to cache list of objects frequently used by the user-interface.
										The values are loaded from the database on first used, and then stored in a Map for fast access on next use.

* SNP-Seek Software Architecture

This diagram summarizes the different software layers and modules for the web-application.

![](uml/snpseekarch.png)


* UML diagrams

The relationships between the components for each modules are illustrated in the following UML diagrams.


Genotype Module
![](uml/genotype.png)

Variety Module
![](uml/variety.png)


Genomics Module
![](uml/genomics.png)


Domain Models 
![](uml/domain.png)

Genotype Query Sequence

![](uml/genotype_query.png)

### Embedded external sites

* JBrowse ([http://jbrowse.org](http://jbrowse.org)) is a browser-based genome browser implemented in javascript. We added some scripts to display the Genotype track to display variants for all varieties within a region. The JBrowse script with the addid codes are in separate bitbucket project.   

* Vista ([http://pipeline.lbl.gov/cgi-bin/gateway2](http://pipeline.lbl.gov/cgi-bin/gateway2)) is a comparative genomics tool with its own viewer. Use used Vista to compare the 5 reference genomes, and the results are viewable within SNP-Seek.


### Get started with development

To start a new module, we created a set of files to start with as templates. Copy these files then rename to the module name.

		blank_module.zul
		_blank_module.zul

The blank_module page uses the BlankModuleFacade defined in the  **org.irri.iric.portal.blank\_module** package. Copy then modify these into a new package using your module name. 

### Who do I talk to? ###

For comments and suggestions, please email us at iric@irri.org
