<html>
	
<head>
	<title>Advance SNP Query</title>
	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="generator" content="HelpNDoc Personal Edition 4.5.0.497">
	<meta name="keywords" content="">
	
	<link type="text/css" rel="stylesheet" media="all" href="css/reset.css" />
	<link type="text/css" rel="stylesheet" media="all" href="css/base.css" />
	<link type="text/css" rel="stylesheet" media="all" href="css/hnd.css" />
	<!--[if lte IE 8]>
		<link type="text/css" rel="stylesheet" media="all" href="css/ielte8.css" />
	<![endif]-->
	<style type="text/css">
		#topic_header
		{
			background-color: #00FF00;
		}
	</style>
	<script type="text/javascript" src="js/jquery.min.js"></script>
	<script type="text/javascript" src="js/hnd.js"></script>
	<script type="text/javascript">
		$(document).ready(function()
		{
			if (top.frames.length == 0)
			{
				var sTopicUrl = top.location.href.substring(top.location.href.lastIndexOf("/") + 1, top.location.href.length);
				top.location.href = "RiceSNP-SeekDatabase.html?" + sTopicUrl;
			}
			else if (top && top.FrameTOC && top.FrameTOC.SelectTocItem)
			{
				top.FrameTOC.SelectTocItem("AdvanceSNPQuery");
			}
		});
	</script>
	
	</head>

<body>

<div id="topic_header">
			<div id="topic_header_content">
				<h1 id="topic_header_text">Advance SNP Query</h1>
				
				<div id="topic_breadcrumb">
					<a href="WebInterface.html">Web Interface</a> &rsaquo;&rsaquo; </div>
				</div>
			<div id="topic_header_nav">
				<a href="WebInterface.html"><img src="img/arrow_up.png" alt="Parent"/></a>
				
				<a href="2Varietiesresult.html"><img src="img/arrow_left.png" alt="Previous"/></a>
				
				<a href="StepsinSNPQuery1.html"><img src="img/arrow_right.png" alt="Next"/></a>
				
				</div>
			<div class="clear"></div>
		</div>
	<div id="topic_content">
		
<p></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps8"><span class="rvts15"><br/></span></p>
<p class="rvps8"><img width="807" height="485" alt="" style="padding : 1px;" src="lib/NewItem29.png"/></p>
<p class="rvps8"><span class="rvts15"><br/></span></p>
<p class="rvps7"><span class="rvts15"><br/></span></p>
<p class="rvps12"><span class="rvts15">The SNP Query interface has two functionalities, to compare alleles between two specified varieties, and to compare all varieties with the reference genome. In both cases the query region can be specified either by chromosomal position, or by gene locus.</span></p>
<h1 class="rvps10"><a name="StepsinSNPQuery"></a>
<span class="rvts0"><span class="rvts17"><br/></span></span></h1>
<h1 class="rvps10"><span class="rvts0"><span class="rvts17">Steps to query SNPs</span></span></h1>
<p class="rvps2"><span class="rvts15"><br/></span></p>
<ol style="text-indent: 0px; margin-left: 24px; list-style-position: outside; list-style-type: decimal;">
<li class="rvps2"><span class="rvts15">Specifiy varieties</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps11" style="margin-left: 0px"><span class="rvts15">To compare alleles between </span><span class="rvts18">two varieties</span><span class="rvts15"> - type the </span><span class="rvts18">Variety</span><span class="rvts15"> names in the two boxes. The text boxes have auto-completion feature so one only needs to type the first few characters then select from the drop-down list.</span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts15">To compare reference with a </span><span class="rvts18">subpopulation</span><span class="rvts15"> - select the subpopulation from the listbox. Some items in a choices is a group of more than one subpopulation, </span><span class="rvts18">all indica</span><span class="rvts15"> (ind1,ind2,ind3,indx), </span><span class="rvts18">all japonica</span><span class="rvts15"> (temp,trop,temp/trop,trop/temp) and &nbsp;</span><span class="rvts18">all varieties</span><span class="rvts15"> </span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts15">To compare reference with a</span><span class="rvts18"> list of varieties</span><span class="rvts15"> - select the variety list name in </span><span class="rvts18">My Variety List</span><span class="rvts15">, which are defined in </span><span class="rvts18">My Lists</span><span class="rvts15">.</span></li>
</ol>
<li class="rvps2"><span class="rvts15">Specify genome region</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps11" style="margin-left: 0px"><span class="rvts15">Select the </span><span class="rvts18">Reference Genome</span><span class="rvts15"> - select among the five sequenced rice genomes to use as reference genome (Japonica Nipponbare, 93-11, IR64-21, DJ123, Kasalath). The chromosome/scaffold and gene locus selections depends on the selected reference genome. </span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts15">To sepecify </span><span class="rvts18">Chromosome/Scaffold region</span><span class="rvts15"> - select the </span><span class="rvts18">chromosome/scaffold</span><span class="rvts15"> name, type the </span><span class="rvts18">start</span><span class="rvts15"> and </span><span class="rvts18">end</span><span class="rvts15"> base positions. Chromosome is avaiable when a fully assembled genome (Nipponbare, Kasalath) is selected, otherwise (IR64-21, DJ123) scaffold is used. 93-11 has both chromosomes and scaffolds.</span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts15">To specify Gene locus - type the </span><span class="rvts18">Gene</span><span class="rvts15"> locus name in the text box. For Nipponbare, we are using the MSU Rice Genome &nbsp;Annotation Project annotations (</span><a class="rvts19" href="http://rice.plantbiology.msu.edu/">http://rice.plantbiology.msu.edu/</a><span class="rvts15">), so gene loci names are of the form LOC_OSXXgYYYYY. &nbsp;For the other 4 reference genomes, we are proposing a new locus naming convention described below. The gene text box also has auto-completion feature. When a locus is selected, the chromosomal position boxes are filled by the selected gene's position. </span></li>
</ol>
<li class="rvps2"><span class="rvts15">Set options</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps11" style="margin-left: 0px"><span class="rvts18">Show all reference alleles</span><span class="rvts15"> – The alleles of all 5 reference genomes is display for each SNP position.</span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts18">Show Nipponbare positions</span><span class="rvts15"> - since the bulk of our SNPs are called against Nipponbare, it is stored in the database using Nipponbare position. When another genome (non-Nipponbare) is used as reference, Nipponbare positions are converted into the basepair coordinate of the selected reference using the genome-wide pairwwise alignment data. Check this option to also display the Nipponbare positions, along with the converted positions.</span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts18">SNP coloring</span><span class="rvts15"> determines the color of alleles as displayed in the table and in the Genome Browser. Options are:</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-roman;">
<li class="rvps13" style="margin-left: 0px"><span class="rvts15">Reference mismatch – SNPs which have polymorphism with the reference are colored red, while the rest is white.</span></li>
<li class="rvps13" style="margin-left: 0px"><span class="rvts15">Nucleotide – SNPs are colored based on the base nucleotide: A green, T red, G orange, C blue. This follows the JBrowse color code.</span></li>
</ol>
<li class="rvps11" style="margin-left: 0px"><span class="rvts18">Core SNPs only</span><span class="rvts15"> -- if checked, will query only the Core SNPs (500,000) instead of the SNP universe (20M). The core set was generated using statistical criteria.</span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts18">Mismatch Only</span><span class="rvts15"> -- if checked, only SNPs having allele mismatches are fetched in the two-variety comparison</span></li>
<li class="rvps11" style="margin-left: 0px"><span class="rvts18">Reset</span><span class="rvts15"> button is provided to clear all input fields.</span></li>
</ol>
<li class="rvps2"><span class="rvts15">Display or Download</span></li>
</ol>
<ol style="text-indent: 0px; margin-left: 24px; list-style-position: outside; list-style-type: decimal;"><ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps9" style="margin-left: 0px"><span class="rvts15">To download the result, click the </span><span class="rvts18">CSV</span><span class="rvts15"> or </span><span class="rvts18">Tab</span><span class="rvts15"> button to get the given format.</span></li>
<li class="rvps9" style="margin-left: 0px"><span class="rvts15">To Display the result in a Table, or view in JBrowse or a Tree, click the </span><span class="rvts18">Search</span><span class="rvts15"> button. </span></li>
</ol></ol>
<p class="rvps9"><span class="rvts15"><br/></span></p>
<p class="rvps7"><a name="AllVarietiesresult"></a>
<span class="rvts15"><br/></span></p>
<h1 class="rvps10"><span class="rvts0"><span class="rvts17">All Varieties result views </span></span></h1>
<p class="rvps12"><span class="rvts15">Three views are available for SNP query results</span></p>
<p class="rvps7"><span class="rvts15"><br/></span></p>
<p class="rvps14"><span class="rvts20">Table</span><span class="rvts16"> &nbsp;- this view displays the Alleles for each variety in a row, for all SNP positions within &nbsp;the specified region or gene. The color is based on SNP coloring selected. The table is displayed in several pages, depending on the number of returned varieties. Note that only varieties having at least one reference mismatch are returned, and the varieties are sorted by decreasing number of allele mismatches. Due to its typlical large size, the result is displayed in several pages and the number of varieties per page depends on the length of the query region.</span></p>
<p class="rvps8"><img width="615" height="331" alt="" style="padding : 1px;" src="lib/NewItem8.png"/></p>
<p class="rvps7"><span class="rvts15"><br/></span></p>
<p class="rvps9"><span class="rvts18">JBrowse</span><span class="rvts15"> – the JBrowse genome browser lays out the SNP genotyping result in the SNP Genotyping track. Each variety is represented as a thin row in the track, and sorted in the same order as in the table. This track is dynamically generated from the query result, and covers only the queried region or locus. Zooming out from the initial region gives empty rows at both sides. To view a wider region, do a query again specifying the new wider region. Unlike in the Table view where results are displayed in several pages, Jbrowse displays all varieties having allele mismatch in the queried region at once.</span></p>
<p class="rvps9"><span class="rvts15"><br/></span></p>
<p class="rvps8"><img width="607" height="318" alt="" style="padding : 1px;" src="lib/NewItem7.png"/></p>
<p class="rvps7"><span class="rvts15"><br/></span></p>
<p class="rvps14"><span class="rvts20">PhyloTree</span><span class="rvts16"> -- Using the number of allele mismatches between varieties within the queried region or gene to compute for distance, the phylogenetic tree is constructed by Neighbor-Joining alorithm. Note that this operation takes some time (~3 minutes for a typical gene length).</span></p>
<p class="rvps14"><span class="rvts16"><br/></span></p>
<h1 class="rvps10"><a name="2Varietiesresult"></a>
<span class="rvts0"><span class="rvts17"><br/></span></span></h1>
<h1 class="rvps10"><span class="rvts0"><span class="rvts17">Two varieties result views</span></span></h1>
<p class="rvps12"><span class="rvts15">For two varieties query, results can be viewed in a table or in JBrowse.</span></p>
<p class="rvps7"><span class="rvts15"><br/></span></p>
<p class="rvps9"><span class="rvts18">Table</span><span class="rvts15"> - when two varieties are compared, the result is a list of SNP position within the specified region. The allele of the reference, and the two varieties for each position are displayed in a row. Polymorphism between the two varieties is highlighted with color red.</span></p>
<p class="rvps8"><img width="534" height="415" alt="" style="padding : 1px;" src="lib/NewItem6.png"/></p>
<p class="rvps7"><span class="rvts15"><br/></span></p>
<p class="rvps9"><span class="rvts18">JBrowse</span><span class="rvts15"> – the JBrowse tab displays the SNPs of the two varieties within the genome, along with the genes. The boxes are colored based on the selected SNP coloring option.</span></p>
<p class="rvps8"><img width="528" height="333" alt="" style="padding : 1px;" src="lib/NewItem5.png"/></p>
<p class="rvps8"><span class="rvts14"><br/></span></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps7"><a name="GeneLocusNaming"></a>
<span class="rvts14"><br/></span></p>
<p class="rvps7"><span class="rvts18">Gene Locus Naming</span></p>
<p class="rvps2"><span class="rvts6"><br/></span></p>
<p class="rvps2"><span class="rvts6">Except for Nipponbare, the four new rice genomes haven't assigned gene locus names yet. The available GFF files for gene annotations still use the gene annotation pipelines' default (like FGENESH, Maker-P). To integrate all these genomes into IRIC, it is therefore necessary that we assign them locus names for uniformity and easy recall. And we propose this scheme:</span></p>
<p class="rvps2"><span class="rvts6"><br/></span></p>
<p class="rvps2"><span class="rvts6">1. Gene name starts with Os. Os stands for Oryza sativa.</span></p>
<p class="rvps2"><span class="rvts6">2. Next 5 characters define the variety:</span></p>
<p class="rvps2"><span class="rvts6"> &nbsp; &nbsp;Nippo -- Nipponbare</span></p>
<p class="rvps2"><span class="rvts6"> &nbsp; &nbsp;9311_ -- 9311</span></p>
<p class="rvps2"><span class="rvts6"> &nbsp; &nbsp;Kasal -- Kasalath</span></p>
<p class="rvps2"><span class="rvts6"> &nbsp; &nbsp;DJ12_ -- DJ123</span></p>
<p class="rvps2"><span class="rvts6"> &nbsp; &nbsp;IR64_ -- IR64</span></p>
<p class="rvps2"><span class="rvts6">3. If there is a chromosome assembly, the next 2 digits are the chromosome number. </span></p>
<p class="rvps2"><span class="rvts6">If there is no chromosome assembly, then the next 5 digits are the contig number. A contig number is a rank of the contigs sorted by length. The longest contig has number 1. If contigs are of the same length, they are sorted by their sequence in alphabetical order.</span></p>
<p class="rvps2"><span class="rvts6">4. The next character is g. g stands for genomic. Instead of g can be m for mitochondrial DNA or c for chloroplast DNA.</span></p>
<p class="rvps2"><span class="rvts6">5. The next 6 digits are the gene number in the sequence with increment of 50. The first gene should have number 010050, the second 010100, the third 010150, etc.</span></p>
<p class="rvps2"><span class="rvts6">Gene names in the chromosomes will be preserved in future versions of annotations. Gene names in scaffolds are for temporary use.</span></p>
<p class="rvps2"><span class="rvts6">6. Alternatively spliced gene models or different versions of the same gene prediction are distinguished by a number added at the end of the gene after the dot. </span></p>
<p class="rvps2"><span class="rvts6"><br/></span></p>
<p class="rvps2"><span class="rvts6">Here are some examples: OsNippo01g210150.1, OsIR64_00342g210150.1, Os9311_11g210150.1, OsDJ12_00342g210150.1</span></p>
<p class="rvps2"><span class="rvts6"><br/></span></p>
<p class="rvps2"><span class="rvts6">We also propose to combine the MSU7 and RAP gene annotations for Nipponbare, and rename the combined gene loci following the scheme above.</span></p>
<p class="rvps2"><span class="rvts6"><br/></span></p>
<p class="rvps2"><span class="rvts6"><br/></span></p>
<p class="rvps2"><span class="rvts6"><br/></span></p>
<p></p>
<p class="rvps6"><span class="rvts12">Created with the Personal Edition of HelpNDoc: </span><a class="rvts13" href="http://www.helpndoc.com/help-authoring-tool">Free help authoring environment</a></p>
</div>
	
	<div id="topic_footer">

			<div id="topic_footer_content">
				Copyright &copy; 2014 by International Rice Informatics Consortium. All Rights Reserved.</div>
		</div>
	</body>
	
</html>

