<html>
	
<head>
	<title>SNP Query</title>
	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="generator" content="HelpNDoc Personal Edition 4.5.0.497">
	<meta name="keywords" content="CSV, Tab, query SNPs, , All Varieties result, Two varieties result, Gene">
	
	<link type="text/css" rel="stylesheet" media="all" href="css/reset.css" />
	<link type="text/css" rel="stylesheet" media="all" href="css/base.css" />
	<link type="text/css" rel="stylesheet" media="all" href="css/hnd.css" />
	<!--[if lte IE 8]>
		<link type="text/css" rel="stylesheet" media="all" href="css/ielte8.css" />
	<![endif]-->
	<style type="text/css">
		#topic_header
		{
			background-color: #00FF00;
		}
	</style>
	<script type="text/javascript" src="js/jquery.min.js"></script>
	<script type="text/javascript" src="js/hnd.js"></script>
	<script type="text/javascript">
		$(document).ready(function()
		{
			if (top.frames.length == 0)
			{
				var sTopicUrl = top.location.href.substring(top.location.href.lastIndexOf("/") + 1, top.location.href.length);
				top.location.href = "RiceSNP-SeekDatabase.html?" + sTopicUrl;
			}
			else if (top && top.FrameTOC && top.FrameTOC.SelectTocItem)
			{
				top.FrameTOC.SelectTocItem("SNPQuery");
			}
		});
	</script>
	
	</head>

<body>

<div id="topic_header">
			<div id="topic_header_content">
				<h1 id="topic_header_text">SNP Query</h1>
				
				<div id="topic_breadcrumb">
					<a href="WebInterface.html">Web Interface</a> &rsaquo;&rsaquo; </div>
				</div>
			<div id="topic_header_nav">
				<a href="WebInterface.html"><img src="img/arrow_up.png" alt="Parent"/></a>
				
				<a href="WebInterface.html"><img src="img/arrow_left.png" alt="Previous"/></a>
				
				<a href="VarietyQuery.html"><img src="img/arrow_right.png" alt="Next"/></a>
				
				</div>
			<div class="clear"></div>
		</div>
	<div id="topic_content">
		
<p></p>
<p class="rvps7"><span class="rvts13"><br/></span></p>
<p class="rvps8"><span class="rvts14"><br/></span></p>
<p class="rvps8"><img width="698" height="370" alt="" style="padding : 1px;" src="lib/NewItem18.png"/></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps13"><span class="rvts14">The SNP Query interface has two functionalities, to compare alleles between two specified varieties, and to compare all varieties with the reference genome. In both cases the query region can be specified either by chromosomal position, or by gene locus.</span></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<h1 class="rvps11"><span class="rvts0"><span class="rvts17">Steps to query SNPs</span></span></h1>
<p class="rvps2"><span class="rvts14"><br/></span></p>
<ol style="text-indent: 0px; margin-left: 24px; list-style-position: outside; list-style-type: decimal;">
<li class="rvps2"><span class="rvts14">Specifiy varieties</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps12" style="margin-left: 0px"><span class="rvts14">To compare alleles between two varieties - type the </span><span class="rvts18">Variety</span><span class="rvts14"> names in the two boxes. The text boxes have auto-completion feature so one only needs to type the first few characters then select from the drop-down list.</span></li>
<li class="rvps12" style="margin-left: 0px"><span class="rvts14">To compare alleles in all varieties - check the </span><span class="rvts18">All varieties</span><span class="rvts14"> checkbox. Varieties selected in the two text boxes are ignored if this is checked.</span></li>
</ol>
<li class="rvps2"><span class="rvts14">Specify genome region</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps12" style="margin-left: 0px"><span class="rvts14">To sepecify chromosome region - select the </span><span class="rvts18">chromosome</span><span class="rvts14"> number, type the </span><span class="rvts18">start</span><span class="rvts14"> and </span><span class="rvts18">end</span><span class="rvts14"> base positions. The interface will checks if the entries are valid.</span></li>
<li class="rvps12" style="margin-left: 0px"><span class="rvts14">To specify Gene locus - type the </span><span class="rvts18">Gene</span><span class="rvts14"> locus name in the text box. We are using the MSU Rice Genome &nbsp;Annotation Project annotations (</span><a class="rvts19" href="http://rice.plantbiology.msu.edu/">http://rice.plantbiology.msu.edu/</a><span class="rvts14">), so gene loci names are of the form LOC_OSXXgYYYYY. &nbsp;The gene text box also has auto-completion feature.</span></li>
</ol>
<li class="rvps2"><span class="rvts14">Set options</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps12" style="margin-left: 0px"><span class="rvts18">SNP coloring</span><span class="rvts14"> determines the color of alleles as displayed in the table and in the Genome Browser. Options are:</span></li>
<ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-roman;">
<li class="rvps14" style="margin-left: 0px"><span class="rvts14">Reference mismatch – SNPs which have polymorphism with the reference are colored red, while the rest is white.</span></li>
<li class="rvps14" style="margin-left: 0px"><span class="rvts14">Nucleotide – SNPs are colored based on the base nucleotide: A green, T red, G orange, C blue. This follows the JBrowse color code.</span></li>
</ol>
<li class="rvps12" style="margin-left: 0px"><span class="rvts18">Reset</span><span class="rvts14"> button is provided to clear all input fields.</span></li>
</ol>
<li class="rvps2"><span class="rvts14">Display or Download</span></li>
</ol>
<ol style="text-indent: 0px; margin-left: 24px; list-style-position: outside; list-style-type: decimal;"><ol style="text-indent: 0px; margin-left: 48px; list-style-position: outside; list-style-type: lower-alpha;">
<li class="rvps10" style="margin-left: 0px"><span class="rvts14">To download the result, click the </span><span class="rvts18">CSV</span><span class="rvts14"> or </span><span class="rvts18">Tab</span><span class="rvts14"> button to get the given format.</span></li>
<li class="rvps10" style="margin-left: 0px"><span class="rvts14">To Display the result in a Table, or view in JBrowse or a Tree, click the </span><span class="rvts18">Search</span><span class="rvts14"> button. </span></li>
</ol></ol>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<h1 class="rvps11"><span class="rvts0"><span class="rvts17">All Varieties result views </span></span></h1>
<p class="rvps13"><span class="rvts14">Three views are available for SNP query results</span></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps15"><span class="rvts15">Table &nbsp;- this view displays the Alleles for each variety in a row, for all SNP positions within &nbsp;the specified region or gene. The color is based on SNP coloring selected. The table is displayed in several pages, depending on the number of returned varieties. Note that only varieties having at least one reference mismatch are returned, and the varieties are sorted by decreasing number of allele mismatches. Due to its typlical large size, the result is displayed in several pages and the number of varieties per page depends on the length of the query region.</span></p>
<p class="rvps8"><img width="615" height="331" alt="" style="padding : 1px;" src="lib/NewItem8.png"/></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps10"><span class="rvts14">JBrowse – the JBrowse genome browser lays out the SNP genotyping result in the SNP Genotyping track. Each variety is represented as a thin row in the track, and sorted in the same order as in the table. This track is dynamically generated from the query result, and covers only the queried region or locus. Zooming out from the initial region gives empty rows at both sides. To view a wider region, do a query again specifying the new wider region. Unlike in the Table view where results are displayed in several pages, Jbrowse displays all varieties having allele mismatch in the queried region at once.</span></p>
<p class="rvps10"><span class="rvts14"><br/></span></p>
<p class="rvps8"><img width="607" height="318" alt="" style="padding : 1px;" src="lib/NewItem7.png"/></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps15"><span class="rvts15">PhyloTree -- Using the number of allele mismatches between varieties within the queried region or gene to compute for distance, the phylogenetic tree is constructed by Neighbor-Joining alorithm. Note that this operation takes some time (~3 minutes for a typical gene length).</span></p>
<p class="rvps9"><span class="rvts16"><br/></span></p>
<h1 class="rvps11"><span class="rvts0"><span class="rvts17">Two varieties result views</span></span></h1>
<p class="rvps13"><span class="rvts14">For two varieties query, results can be viewed in a table or in JBrowse.</span></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps10"><span class="rvts14">Table - when two varieties are compared, the result is a list of SNP position within the specified region. The allele of the reference, and the two varieties for each position are displayed in a row. Polymorphism between the two varieties is highlighted with color red.</span></p>
<p class="rvps8"><img width="534" height="415" alt="" style="padding : 1px;" src="lib/NewItem6.png"/></p>
<p class="rvps7"><span class="rvts14"><br/></span></p>
<p class="rvps10"><span class="rvts14">JBrowse – the JBrowse tab displays the SNPs of the two varieties within the genome, along with the genes. The boxes are colored based on the selected SNP coloring option.</span></p>
<p class="rvps8"><img width="528" height="333" alt="" style="padding : 1px;" src="lib/NewItem5.png"/></p>
<p class="rvps7"><span class="rvts13"><br/></span></p>
<p></p>

</div>
	
	<div id="topic_footer">

			<div id="topic_footer_content">
				Copyright &copy; 2014 by International Rice Informatics Consortium. All Rights Reserved.</div>
		</div>
	</body>
	
</html>

