<!DOCTYPE html>
<html>
<head>
<title>About</title>

<meta name="keywords" content="keyword1,keyword2,keyword3">
<meta name="description" content="this is my page">
<meta name="content-type" content="text/html; charset=ISO-8859-1">

<!--<link rel="stylesheet" type="text/css" href="./styles.css">-->

</head>

<body>

	<div style="font-family: sans-serif; color: darkgreen" align="center">
		<br> <br> <br>


		<table>
			<tr>
				<td width="15%"></td>
				<td align="center">
					<h2>About this site</h2>
				</td>
				<td width="15%"></td>
			</tr>
			<tr>
				<td width="15%"></td>
				<td>
					<p>
						This site provides Genotype, Phenotype, and Variety Information
						for rice (Oryza sativa L.). SNP genotyping data (called against <a
							target="_blank"
							href="http://rice.plantbiology.msu.edu/annotation_pseudo_current.shtml">
							Nipponbare reference Os-Nipponbare-Reference-IRGSP-1.0</a>) came from
						<a target="_blank"
							href="http://www.gigasciencejournal.com/content/3/1/7"> 3,000
							Rice Genomes Project </a>. Phenotype and passport data for the 3,000
						rice varieties came from the International Rice Genebank
						Collection Information System (<a target="_blank"
							href="http://irgcis.irri.org:81/grc/irgcishome.html">IRGCIS</a>).
						We are a part of an ongoing effort by the International Rice
						Informatics Consortium (<a target="_blank"
							href="http://iric.irri.org">IRIC</a>) to centralize information
						access to rice research data and provide computational tools to
						facilitate rice improvement via discovery of new gene-trait
						associations and accelerated breeding.
					</p>

				</td>
				<td width="15%"></td>
			</tr>

			<tr>
				<td width="15%"></td>
				<td align="center">
					<h2>How to Cite SNP-Seek</h2>
				</td>
				<td width="15%"></td>
			</tr>
			<tr>
				<td width="15%"></td>
				<td>

					<h4>
						<i>Primary publication</i>
					</h4> Mansueto, et al. <b>Rice SNP-seek database update: new SNPs,
						indels, and queries</b>.<i>Nucl. Acids Res.</i>(2017) 45 (D1):
					D1075-D1081. <BR> doi: <a target="_pub"
					href="http://nar.oxfordjournals.org/content/early/2016/11/29/nar.gkw1135.abstract">10.1093/nar/gkw1135</a><BR>
					This paper describes new features and datasets added to SNP-Seek in
					2015-2017 as well as software and database updates. </br>
				</td>
				<td width="15%"></td>
			</tr>

			<tr>
				<td width="15%"></td>
				<td><BR></td>
				<td width="15%"></td>
			</tr>

			<tr>
				<td width="15%"></td>
				<td>
					<h4>
						<i>Other publications</i>
					</h4> Mansueto, et al. <b>SNP-Seek II: A resource for allele mining
						and analysis of big genomic data in Oryza sativa</b>.<i>Current
						Plant Biology</i>(2016). 7-8:16-25<BR> doi: <a target="_pub"
					href="http://doi.org/10.1016/j.cpb.2016.12.003">10.1016/j.cpb.2016.12.003</a><BR>
					This paper contains details on variant calling for 5 references,
					integration of additional genomic data, web interface, database
					schema, use cases, web services API. <BR> <BR>
				</td>
				<td width="15%"></td>
			</tr>


			<tr>
				<td width="15%"></td>
				<td>Alexandrov, et al. <b>SNP-Seek database of SNPs derived
						from 3000 rice genomes</b>.<i>Nucl. Acids Res.</i>(2015).
					43(D1):D1023-D1027<BR> doi: <a target="_pub"
					href="http://doi.org/10.1093/nar/gku1039">10.1093/nar/gku1039</a><BR>
					This is the first publication introducing SNP-Seek database, data
					generation for Nipponbare reference,and common use cases. <BR>
					<BR>
				</td>
				<td width="15%"></td>
			</tr>

			<tr>
				<td width="15%"></td>
				<td>Wang W, Mauleon R, Hu Z, Chebotarov D, Tai S, et al. <b>Genomic
						variation in 3,010 diverse accessions of Asian cultivated rice.</b>. <i>Nature.</i>(2018).
					557:43-49<BR> doi: <a target="_pub"
					href="https://www.nature.com/articles/s41586-018-0063-9">10.1038/s41586-018-0063-9</a><BR>
					The main scientific paper for the 3K RGP.<BR> <BR>
				</td>
				<td width="15%"></td>
			</tr>

			<tr>
				<td width="15%"></td>
				<td align="center">
					<h2>Updates log</h2>
				</td>
				<td width="15%"></td>
			</tr>

			<tr>
				<td width="15%"></td>
				<td>
					<h4>
						<i>3/24/2020</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>FIXED: NullPointerException SNPQueryController.java:4702</li>
						<li>FIXED: Removed selection since organism is only limited
							to Nipponbarre. Error thrown is java.io.FileNotFoundException:
							cvterms_biological_process_16.tsv</li>
						<li>FIXED: Error encountered when no workflow is selected.
							Error occurred in generatePlinkfiles method from
							DefaultGalaxyController.java. This is under the galaxy tab in the
							genotype search</li>
					</ul>
					<h4>
						<i>3/17/2020</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: A docker image containing 3kFiltered dataset is
							deployed publicly in docker hub. This image is built on top of
							postgresql docker image.</li>
					</ul>
					<h4>
						<i>2/13/2020</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Deployed SNP-seek(version 5.4.0) with the ff. updates</li>
						<li>UPDATE: Removed "all" option of variety selection in
							download section</li>
						<li>ADDED: ZK Charts license Agreement for development</li>
					</ul>
					<h4>
						<i>1/27/2020</i>
					</h4>
					<ul style="list-style-type: disc">

						<li>BUG FIXED: "Add Filter" in genotype results</li>
					</ul>
					<h4>
						<i>12/18/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>ADDED: "Open Source License Agreement" file in the
							project</li>
						<li>ADDED: A pop-up disclaimer message before downloading a
							sequence.</li>
						<li>BUG FIXED: URL object in galaxyFacade class that is use
							to refer to downloads in galaxy.</li>
					</ul>
					<h4>
						<i>11/12/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>BUG FIXED: NullPointer Exception in querying Genotype and
							Galaxy download</li>
					</ul>

					<h4>
						<i>11/12/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>BUG FIXED: Haplotype view</li>
						<li>BUG FIXED: errors in handling jobs for download.</li>

					</ul>
					<h4>
						<i>10/28/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Removed unused files</li>
						<li>ADDED: License file to Development Docs and in the menu</li>
						<li>ADDED: New folder for Documentations</li>
						<li>REMOVED: ZK charts in project that is publicly accessed
							or for redistribution purposes. Requiring external developers to
							get a license for ZK charts if they want to include charts.</li>
					</ul>

					<h4>
						<i>10/21/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Changed link of "Report Bugs" page</li>
					</ul>
					<h4>
						<i>10/01/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>BUG FIXED: searched genotype using Locus MyList</li>
						<li>UPDATE: beanstalk.properties - changed beanstalk config
							to use https values</li>
						<li>UPDATE: VSnpeffDAOImpl, SnpsStringHDF5nRDBMSHybridService
							- SNP_EFF update</li>
						<li>BUG FIXED: VariantAlignmentTableArraysImpl - change to
							false.. genotype search using NONSYNONYMOUS and a list from user</li>

					</ul>

					<h4>
						<i>09/05/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>NEW: SNP-seek is now using HTTPS.</li>
						<li>UPDATE: Improve SNP effect tab user interface</li>
						<li>UPDATE: Chromosomes now sorted</li>
						<li>BUG FIXED: Error when querying genotype directly using
							non-synonymous option</li>
						<li>BUG FIXED: Empty MDS plot view in the results of Genotype
							Query</li>
						<li>BUG FIXED: Empty SNP effect table.</li>
						<li>BUG FIXED: Haplotype view error</li>
						<li>BUG FIXED: Download CSV file does not have Snp effect
							info</li>
						<li>BUG FIXED: Bug regarding downloading a big dataset</li>
						<li>BUG FIXED: Displaying of manhattan plot in GWAS tab</li>
						<li>BUG FIXED: Bugs in download page caused by the merging.</li>
						<li>BUG FIXED: Bugs in "Add Filter" feature in SNP genotype
							page (SNP seek)</li>
					</ul>

					<h4>
						<i>06/03/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Added Disclaimer note in the contactUS page</li>
						<li>BUG FIXED: Manhattan plot not displaying due to
							NullException when selecting SNP -log list</li>
					</ul>

					<h4>
						<i>05/20/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>BUG FIXED: Resolve upload in MyList by updating jars
							used.</li>
						<li>UPDATE: Merge remote-tracking branch
							'origin/temp-with-galaxy' into SnpSeekMavenProj. Resolved
							conflicting jars</li>
					</ul>

					<h4>
						<i>05/16/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Added exttab, galaxy pages. Exttab moved the
							tabpanels into a separate zul, allowing to be displayed in new
							window on right click of tabs. Galaxy submit long running
							computations into a galaxy instance using blend4j, bioblend and
							jython.Uses python library included in WEB-INF/lib/Lib</li>
						<li>UPDATE: right clickable tabs, Galaxy tab, vcf2fasta in
							galaxy-dev. Right clickable tabs, Galaxy tab, vcf2fasta in
							galaxy-dev http://13.250.4.164:8080</li>
					</ul>

					<h4>
						<i>05/07/2019</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Change naming convention in property files</li>
						<li>UPDATE: Dockerfile to use updated docker image for the
							portal</li>
						<li>BUG FIXED: NullPointerException due to adding of filters</li>
						<li>ADDED: "dataset"" Attribute to
							VariantAlignmentTableArraysImpl class</li>

					</ul>


					<h4>
						<i>12/05/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Database schema has changed. ADDED: 'dataset'
							(String) in SnpAllVars object to make it more unique aside from
							retrieving VarId alone</li>
						<li>UPDATE: Changed Mapping Map(BigDecimal, StockSample) to
							Map(String, Map(BigDecimal, StockSample)) to display multiple
							dataset with similar samples</li>
						<li>UPDATE: ListItemsDAOAllImpl - remove if statement that
							handles one dataset. instead implemented it by retrieval by Set
							regardless how many dataset.</li>
						<li>UPDATE: Bug Fixed in displaying multiple dataset with
							similar samples</li>
						<li>ADDED: Dataset variable to IndelStringAllVarsImpl class
							and a new Mapping variable Variety to Dataset variable.</li>
						</li>


					</ul>

					<h4>
						<i>11/06/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Project converted to a Maven project</li>
					</ul>

					<h4>
						<i>10/16/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>ADDED: Legacy Web Services and usecases. GetAllCOTerms,
							GetCOTerm, GetAllCOTerms given a variety, Get Record by Variety
							and CO Term</li>
						<li>FIXED: "Download Error Message" in bulk download</li>
					</ul>

					<h4>
						<i>10/10/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>FIXED: MDS PLOT error.
							snp-seek.irri.org/_variety.zul?irisid=allmds returns error 500</li>
					</ul>

					<h4>
						<i>09/11/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>ADDED: "OS" tag to config file for setting values for
							OPERATING SYSTEM. This is to determine what HDF5 libraries are to
							be loaded.</li>
						<li>ADDED: defaultOrganism in the webserver config file to
							set initial value for organism.</li>
						<li>FIXED: Nullpointer for BigListBox in SNP Genotype tab.
						<li>ADDED: dockerfile. This dockerfile deploys precompiled
							JAVA codes of snpseek portal to docker.</li>
					</ul>

					<h4>
						<i>09/11/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>FIXED: NULL Pointer Exception blast servername.</li>
						<li>ADDED: filter for chr String, filtering whole word
							"Chromosome"</li>
						<li>UPDATE: Vscaddolds organism to retrieve name in getName
							method instead of getting the uniquename</li>
						<li>FIXED: "Processing icon" that locks the screen even an
							exception was thrown</li>
						<li>FIXED: Null pointer exception in GenotypeWS</li>
						<li>ADDED: New citation in the main page.</li>
						<li>ADDED: Info in the page regarding why other reference
							genomes are missing</li>
					</ul>

					<h4>
						<i>08/01/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Home Page. Added "How to Cite SNP seek.</li>
						<li>Fixed: BAM and VCF Download in the Download Page</li>
					</ul>

					<h4>
						<i>07/16/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>UPDATE: Publication page content. Page format of this
							page was also changed.</li>
						<li>FIXED: BrAPI web services VAllStockBasicpropDaoImpl.</li>
						<li>ADDED: findVarietyByNamesLike and findVarietyById</li>
						<li>FIXED: findVarietiesByDataset. If all is selected, All
							varAllStock.</li>
					</ul>

					<h4>
						<i>07/05/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>BrAPI web services: Fix applied to usecases like
							germplasmPUI, findVarietyByNamesLike when input dataset is set to
							"all".</li>
						<li>BrAPI web services: Implemented usecases
							findVAllstockBasicpropByNamesLikeDataset and
							findAllVAllstockBasicprops.</li>
					</ul>

					<h4>
						<i>06/20/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Added option to phenotype list (in Variety Tab) to viewed
							in legacy terms or as CO Terms.</li>
						<li>Removed case sensitivity to List Combo boxes e.g.
							Chromosome, Locus in Genotype Search and Sequence download page.
							And Designation, Accession, Assay Id in variety search.</li>
					</ul>

					<h4>
						<i>05/16/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Deployed new user-interface for SNP-Seek.</li>
						<li>Added "download as CSV" button to download GWAS results
							in CSV file format.</li>
						<li>Added Confirmation message after a successful creation of
							user defined list in Variety Search.</li>
						<li>Implemented findAllVariety which accepts a set of dataset
							and return a set of varieties.</li>
					</ul>

					<h4>
						<i>02/03/2018</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Implemented an email service using AWS SES mailer to send
							seed orders.</li>
						<li>Set SNP Seek portal to use configuration files to toggle
							portal components' visibility.</li>
						<li>Developed new schema to easily add new datasets.</li>
						<li>Independent selection of dataset and SNP set.</li>
					</ul>

					<h4>
						<i>12/01/2017</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Added page for seed orders from IRRI Genebank.</li>
					</ul>

					<h4>
						<i>08/01/2017</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>New feature to cluster genotypes and display in haplotype
							viewer.</li>
					</ul>

					<h4>
						<i>03/06/2017</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>RDBMS migrated to PostgreSQL in AWS-RDS</li>
						<li>Trait genes browser (all chromosomes viewer/ideogram)</li>
						<li>Four (MH63,ZS97,N22,IR8) new reference
							genomes/cdna/cds/pep made available for BLAST search</li>
					</ul>

					<h4>
						<i>12/12/2016</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Query gene loci by SNP list annotates the markers and
							genes from various sources</li>
						<li>Genotype similary search with any of the varieties</li>
						<li>Search and highlight variety or variety list in MDS plot</li>
						<li>Make MDS points proportional to phenotype value</li>
						<li>Submit locus list/gene set to enrichment/network analysis
							sites</li>
					</ul>

					<h4>
						<i>09/19/2016</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>URL changed to http://snp-seek.irri.org</li>
						<li>HDRA varieties added with phenotype data from GRIMS</li>
						<li>Option to use either 3kRGP, HDRA or both datasets</li>
						<li>Web services documentation made public</li>
						<li>Alternate sequence download using GATK tool</li>
					</ul>

					<h4>
						<i>05/16/2016</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Option to query SNP set to either (All 32M SNPs in
							previous version, Base 18M, Filtered 4.8M or Core 404k)</li>
						<li>Subpopulations renamed: trop1->trop, trop2->subtrop</li>
						<li>Result for SNP Effect program on MSU7 gene loci</li>
						<li>More SNP annotations in Locus query by SNP List
							(promoters, RiceNet and PRIN gene interactions)</li>
					</ul>

					<h4>
						<i>03/23/2016</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Increase QNP query length to 5mb region, or lists with
							1000 gene loci, or 500k SNP positions (not displayed but for
							download only)</li>
						<li>Query and match the 3k varieties using genotypes (SNP
							list with allele column)</li>
						<li>Matching option for missing alleles</li>
						<li>Long downloads are run in the background</li>
					</ul>

					<h4>
						<i>02/04/2016</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Search genes using gene names/symbols/accessions</li>
						<li>Added TO/PO ontology (from Oryzabase) and QTARO trait
							genes annotation to SNP lists</li>
						<li>Gene information page and external links</li>
					</ul>

					<h4>
						<i>12/30/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>New homepage icons</li>
						<li>Support for multiple reference genomes (Nipponbare, 9311,
							IR64-21, Kasalath, DJ123)</li>
						<li>Display alleles in all 5 reference genomes in SNP query</li>
						<li>Display other reference genomes and gene models in
							JBrowse</li>
						<li>Display pairwise alignments between any of the 5
							reference genomes</li>
						<li>Variety phenotypes are mapped and can be queried by Rice
							Crop Ontology (C0_320) and Plant trait Ontology (TO)</li>
						<li>Locus query for SNP lists returns genes for 3 Nipponbare
							gene models, QTARO and Gramene QTLs</li>
					</ul>

					<h4>
						<i>10/26/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>In addition to MSU7, gene loci now includes RAP and our
							merged MSU7, RAP and FGenesh++ gene models. Loci from any of the
							three are merged if they have >50% overlap on the CDS</li>
						<li>Query gene loci by Gene Ontology terms</li>
						<li>Query gene loci by Traits (QTARO) and Trait and Plant
							Ontologies (from TO, PO using Oryzabase)</li>
					</ul>


					<h4>
						<i>09/29/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Using updated SNPs/Indel data with 3024 varieties</li>
						<li>Updated subpopulations</li>
						<li>Download links for BAM and VCF from Amazon S3</li>
						<li>New JBrowse tracks for BAM, VCF, Trait loci, Gene models</li>
						<li>Option to include Indels in query</li>
						<li>Option to display phenotype value in genotype table</li>
						<li>Generate alternate sequence from the VCF files</li>
					</ul>

					<h4>
						<i>06/02/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Download links for FastQ and FastQC reports</li>
						<li>Allele frequency calculation on SNP Query results</li>
					</ul>

					<h4>
						<i>05/04/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>SNP Query using SNP list with multiple
							contigs/chromosomes in a list</li>
						<li>SNP Query using Gene Loci list</li>
						<li>Transitive closure is used for Loci search by Gene
							Ontology term</li>
						<li>Added Loci search by sequence using BLAST on locus, cds,
							protein and 3kbp-upstream sequences for all 5 reference genomes</li>
					</ul>

					<h4>
						<i>04/13/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Allow filtering of genotype table by ALLELE value,
							variety and mismatch</li>
						<li>Allow sorting of genotype table by ALLELE value, variety
							and mismatch</li>
						<li>SNP Genotype track in JBrowse is updated after filtering
							and sorting operations</li>
						<li>Gene loci search by SNP list supports multiple
							contigs/chromosomes in a list</li>
					</ul>

					<h4>
						<i>03/30/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Added search Gene loci by region and SNP list</li>
						<li>Added search Gene loci by sequence using Blast on
							Nipponbare cds, cdna and proteins</li>
						<li>Added search Genomic region by sequence using Blast on
							Nipponbare, IR64-21, 9311, DJ123 and Kasalat genomes</li>
					</ul>

					<h4>
						<i>03/12/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Added search Gene loci by gene function</li>
						<li>Create gene locus list from Loci Search result and
							through My Lists editor</li>
					</ul>

					<h4>
						<i>03/3/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Added options to include, exclude or highlight
							non-synonymous/synonymous SNPs</li>
						<li>Highlight slice acceptor and splice donor SNPs</li>
						<li>Query length for Core SNPs limited to 1Mbp, and All SNPs
							to 50kbp to avoid Out of Memory Error and crashing the server</li>
					</ul>

					<h4>
						<i>02/16/2015</i>
					</h4>
					<ul style="list-style-type: disc">
						<li>Country of origin is updated (from GRIMS passport field
							ori_country to country_origin)</li>
						<li>Added option to Include, Highlight or Exclude Synonymous
							SNPs</li>
						<li>Genotype table is revised from ZK Grid to use ZK
							Biglistbox for fast rendering of long regions</li>
						<li>Added Alignment Tab panel to display, sort, or export
							Genotype data as Multiple Sequence Alignments (in preparation to
							display Indels)</li>
					</ul>




				</td>
				<td width="15%"></td>
			</tr>



		</table>
		<BR> <BR> <BR> <BR>
	</div>





</body>
</html>
