<!DOCTYPE html>
<html>
  <head>
    <title>About</title>
	
    <meta name="keywords" content="keyword1,keyword2,keyword3">
    <meta name="description" content="this is my page">
    <meta name="content-type" content="text/html; charset=ISO-8859-1">
    
    <!--<link rel="stylesheet" type="text/css" href="./styles.css">-->

  </head>
  
  <body>
   
   <div style="font-family:sans-serif;color:darkgreen" align="center">
   <br><br><br><br><br><br><br>
  <h3>
  About this site
  </h3>
  <br><br>

<table>
<tr>
<td width="30%"></td><td>   

This site provides Genotype, Phenotype, and Variety Information for rice (Oryza sativa L.). SNP genotyping data (called against 
 <a target="_blank" href="http://rice.plantbiology.msu.edu/annotation_pseudo_current.shtml">Nipponbare reference Os-Nipponbare-Reference-IRGSP-1.0</a>) came from <a target="_blank" href="http://www.gigasciencejournal.com/content/3/1/7"> 3,000 Rice Genomes Project </a>. 
Phenotype and passport data for the 3,000 rice varieties came from the International Rice Information System (<a target="_blank" href="http://iris.irri.org">IRIS</a>).
We are a part of an ongoing effort by the International Rice Informatics Consortium (<a target="_blank" href="http://iric.irri.org">IRIC</a>) to centralize information access to rice research data and provide computational tools to facilitate rice improvement via discovery of new gene-trait associations and accelerated breeding. 


 </td>
 <td width="30%"></td>
 </tr>
 </table>


<h3>Updates log</h3>
 
</div>


<table>
<tr>
<td width="10%"></td><td>   


<h4>2/16/2015</h4>
<ul style="list-style-type:disc">
 <li>Country of origin is updated (from GRIMS passport field ori_country to country_origin)</li>
 <li>Added option to Include, Highlight or Exclude Synonymous SNPs </li>
 <li>Genotype table is revised from ZK Grid to use ZK Biglistbox for fast rendering of long regions</li>
 <li>Added Alignment Tab panel to display, sort, or export Genotype data as Multiple Sequence Alignments (in preparation to display Indels)</li> 
</ul>

<h4>3/3/2015</h4>
<ul style="list-style-type:disc">
 <li>Added options to include, exclude or highlight non-synonymous/synonymous SNPs</li>
 <li>Highlight slice acceptor and splice donor SNPs</li>
 <li>Query length for Core SNPs limited to 1Mbp, and All SNPs to 50kbp to avoid Out of Memory Error and crashing the server</li>
 </ul>

<h4>3/12/2015</h4>
<ul style="list-style-type:disc">
 <li>Added search Gene loci by gene function</li>
 <li>Create gene locus list from Loci Search result and through My Lists editor</li>
</ul>

<h4>3/30/2015</h4>
<ul style="list-style-type:disc">
 <li>Added search Gene loci by region and SNP list</li>
 <li>Added search Gene loci by sequence using Blast on Nipponbare cds, cdna and proteins</li>
 <li>Added search Genomic region by sequence using Blast on Nipponbare, IR64-21, 9311, DJ123 and Kasalat genomes</li>
</ul>


<h4>4/13/2015</h4>
<ul style="list-style-type:disc">
 <li>Allow filtering of genotype table by ALLELE value, variety and mismatch</li>
 <li>Allow sorting of genotype table by ALLELE value, variety and mismatch</li>
 <li>SNP Genotype track in JBrowse is updated after filtering and sorting operations</li>
 <li>Gene loci search by SNP list supports multiple contigs/chromosomes in a list</li>
 </ul>

<h4>5/04/2015</h4>
<ul style="list-style-type:disc">
 <li>SNP Query using SNP list with multiple contigs/chromosomes in a list</li>
 <li>SNP Query using Gene Loci list</li>
 <li>Transitive closure is used for Loci search by Gene Ontology term</li>
 <li>Added Loci search by sequence using BLAST on locus, cds, protein and 3kbp-upstream sequences for all 5 reference genomes</li>
 </ul>

<h4>6/02/2015</h4>
<ul style="list-style-type:disc">
 <li>Download links for FastQ and FastQC reports</li>
 <li>Allele frequency calculation on SNP Query results</li>
 </ul>

<h4>9/29/2015</h4>
<ul style="list-style-type:disc">
 <li>Using updated SNPs/Indel data with 3024 varieties</li>
 <li>Updated subpopulations</li>
 <li>Download links for BAM and VCF from Amazon S3</li>
 <li>New JBrowse tracks for BAM, VCF, Trait loci, Gene models</li>
 <li>Option to include Indels in query</li>
 <li>Option to display phenotype value in genotype table</li>
 <li>Generate alternate sequence from the VCF files</li>
 </ul>
 
<h4>10/26/2015</h4>
<ul style="list-style-type:disc">
 <li>In addition to MSU7, gene loci now includes RAP and our merged MSU7, RAP and FGenesh++ gene models</li>
 <li>Query gene loci by Gene Ontology terms</li>
 <li>Query gene loci by Traits (QTARO) and Trait and Plant Ontologies (from TO, PO using Oryzabase)</li>
 </ul>


  
</td>
 <td width="10%"></td>
 </tr>
 </table>
 
  </body>
</html>
